source/common/quic/platform/flags_impl.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+#include <string>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/synchronization/mutex.h"
+
+namespace quiche {
+
+class Flag;
+
+// Registry of QUICHE flags. Can be used to reset all flags to default values,
+// and to look up and set flags by name.
+class FlagRegistry {
+public:
+  ~FlagRegistry() = default;
+
+  // Return singleton instance.
+  static FlagRegistry& getInstance();
+
+  // Reset all registered flags to their default values.
+  void resetFlags() const;
+
+  // Look up a flag by name.
+  Flag* findFlag(absl::string_view name) const;
+
+private:
+  FlagRegistry();
+
+  const absl::flat_hash_map<std::string, Flag*> flags_;
+};
+
+// Abstract class for QUICHE protocol and feature flags.
+class Flag {
+public:
+  // Construct Flag with the given name and help string.
+  Flag(const char* name, const char* help) : name_(name), help_(help) {}
+  virtual ~Flag() = default;
+
+  // Set flag value from given string, returning true iff successful.
+  virtual bool setValueFromString(const std::string& value_str) = 0;
+
+  // Reset flag to default value.
+  virtual void resetValue() = 0;
+
+  // Return flag name.
+  std::string name() const { return name_; }
```

Comment:
nit: I wonder if this could return a const& or a string_view to avoid copying the underlying string? (ditto below)

Reply:
Great suggestions! done.